src/audio/module_adapter/module/cadence.c

```diff
@@ -207,12 +251,95 @@ static int cadence_codec_post_init(struct processing_module *mod)
 	return 0;
 }
 
+#if CONFIG_IPC_MAJOR_4
 static int cadence_codec_init(struct processing_module *mod)
 {
-	int ret;
+	const struct ipc4_cadence_module_cfg *cfg;
+	struct module_data *codec = &mod->priv;
+	struct cadence_codec_data *cd = NULL;
+	struct module_config *setup_cfg;
 	struct comp_dev *dev = mod->dev;
+	int ret;
+
+	comp_dbg(dev, "cadence_codec_init() start");
+
+	cd = rballoc(0, SOF_MEM_CAPS_RAM, sizeof(struct cadence_codec_data));
+	if (!cd) {
+		comp_err(dev, "cadence_codec_init(): failed to allocate memory for cadence codec data");
+		return -ENOMEM;
+	}
+
+	codec->private = cd;
+	codec->mpd.init_done = 0;
+	cd->self = NULL;
+	cd->mem_tabs = NULL;
+	cd->api = NULL;
+	cd->setup_cfg.avail = false;
+
+	/* copy the setup config only for the first init */
+	if (codec->state == MODULE_DISABLED && codec->cfg.avail) {
+		setup_cfg = &cd->setup_cfg;
+
+		cfg = (const struct ipc4_cadence_module_cfg *)codec->cfg.init_data;
+
+		/* allocate memory for set up config */
+		setup_cfg->data = rballoc(0, SOF_MEM_CAPS_RAM, cfg->param_size);
+		if (!setup_cfg->data) {
+			comp_err(dev, "cadence_codec_init(): failed to alloc setup config");
+			ret = -ENOMEM;
+			goto free;
+		}
+
+		/* allocate memory for runtime set up config */
+		codec->cfg.data = rballoc(0, SOF_MEM_CAPS_RAM, cfg->param_size);
```

Comment:
why `rballoc()` here and in line 286 above? Please review allocation in all the code